### PR TITLE
add 'var' keyword. 

### LINKED ISSUE #197: parser: invalid parsing of if expr
Parser of `if` isn't conforming to specification.

The code below fail to parse:

```sh
fn checkTool(name) {
	toolpath <= which $name

	if $toolpath != $GOPATH+"/bin/"+$name {
		errmsg <= printf "Tool '%s' isn't from github.com/c0defellas/enzo" $name
		return $errmsg
	}

	return ""
}
```

With error below:

```
nashfmt:6:24: Expected '{' but found +
```

### PR DESCRIPTION
**Pull request by tiago4orion**
This PR adds support for the `var` keyword.

The rationale is that currently there's no way to update global variables or update variables in the parent scope. Instead of adding a `global` keyword (python, C, etc) or a `local` keyword (lua), we chose to add a `var` keyword to initialize variables with some value and simple assignments only updating existing variables (Like Go, javascript* and so on).

See example below:

```js
fn counter() {
    var val = "0"
    fn inc() {
        val <= expr $val "+" 1
        return $val
    }
    return $inc
}

var c <= counter()
print("%s\n", c())
print("%s\n", c())
print("%s\n", c())
```

This PR also fix #197 and adds support for function invocation in string concatenation.

### CODE REVIEW

File: internal/sh/shell.go [hunk 1]
```diff
@@ -605,13 +631,32 @@ func (shell *Shell) evalConcat(path ast.Expr) (string, error) {
...
 		case ast.NodeStringExpr:
 			str, ok := part.(*ast.StringExpr)
-
 			if !ok {
 				return "", errors.NewEvalError(shell.filename, part,
 					"Failed to eval string: %s", part)
 			}
 
 			pathStr += str.Value()
+		case ast.NodeFnInv:
```
Comment (tiago4orion) [inline internal/sh/shell.go]: Now we can invoke functions in concatenation:  `"author: "+name()+surname()+"."`
```

File: internal/sh/shell.go [hunk 2]
```diff
@@ -1683,37 +1731,74 @@ func (shell *Shell) evalExpr(expr ast.Expr) (sh.Obj, error) {
 		expr, "Failed to eval expression: %+v", expr)
 }
 
+func (shell *Shell) executeSetenvAssign(assign *ast.AssignNode) error {
+	for i := 0; i < len(assign.Names); i++ {
+		name := assign.Names[i]
+		value := assign.Values[i]
+		err := shell.initVar(name, value)
```
Comment (tiago4orion) [inline internal/sh/shell.go]:
```
setenv PATH = "/bin:/usr/bin"
```
now is like:
```
var PATH = "/bin:/usr/bin"
setenv PATH
```
but do not fail if PATH is already set.

File: internal/sh/shell.go [hunk 3]
```diff
@@ -1742,97 +1827,119 @@ func (shell *Shell) concatElements(expr *ast.ConcatExpr) (string, error) {
...
+	outbuf, errbuf, status, cmdErr := shell.execCmdOutput(cmd, false)
+
+	// Only getting command output
+	// In this case the script must abort in case of errors
 	if len(assign.Names) == 1 {
-		output, status, cmdErr := shell.execCmdOutput(cmd, false)
+		shell.stderr.Write(errbuf) // flush stderr
+		err := shell.setvar(assign.Names[0], sh.NewStrObj(string(outbuf)), local)
 
-		// compatibility mode
```
Comment (tiago4orion) [inline internal/sh/shell.go]: the magical `status` variable was removed.

File: internal/sh/shell.go (hunk 2 above)
Comment (vitorarins) [inline internal/sh/shell.go]: Hmm, this is kind of a bummer. It was nice to set envvars in one liner. But is not the end of the world.

Comment (tiago4orion) [inline internal/sh/shell.go]: Sorry, what I mean was that:

*now when you write:*
`setenv PATH = "/bin:/usr/bin"`
what happens in the interpreter is the same as:
```var PATH = "/bin:/usr/bin"
setenv PATH```
In other words, it always sets a local scoped variable called PATH and set it in the environment (setenv).

File: internal/sh/shell.go [hunk 4]
```diff
@@ -1742,97 +1822,120 @@ func (shell *Shell) concatElements(expr *ast.ConcatExpr) (string, error) {
...
 }
 
-func (shell *Shell) executeExecAssignCmd(v ast.Node) error {
+func (shell *Shell) executeExecAssignCmd(v ast.Node, where bool) error {
 	assign := v.(*ast.ExecAssignNode)
 	cmd := assign.Command()
 
+	// Only getting command output
+	// In this case the script must abort in case of errors
 	if len(assign.Names) == 1 {
```
Comment (tiago4orion) [inline internal/sh/shell.go]: @vitorarins 

```a <= cmd```

File: internal/sh/shell.go [hunk 5]
```diff
@@ -1742,97 +1822,120 @@ func (shell *Shell) concatElements(expr *ast.ConcatExpr) (string, error) {
...
+		return nil
 	}
 
-	if len(assign.Names) != 2 {
-		return errors.NewEvalError(shell.filename,
-			v, "multiple assignment of commands requires two variable names, but got %d",
-			len(assign.Names))
-	}
+	// Only getting stdout and exit status
+	if len(assign.Names) == 2 {
```
Comment (tiago4orion) [inline internal/sh/shell.go]: @vitorarins 

```out, status <= cmd```

File: internal/sh/shell.go [hunk 6]
```diff
@@ -1742,97 +1822,120 @@ func (shell *Shell) concatElements(expr *ast.ConcatExpr) (string, error) {
...
+			return err
+		}
 
-	if err != nil {
-		return err
+		return nil
 	}
 
-	err = shell.setvar(assign.Names[1], status)
+	if len(assign.Names) == 3 {
```
Comment (tiago4orion) [inline internal/sh/shell.go]: @vitorarins 

```
outbuf, errbuf, status <= cmd
```

**Comment by tiago4orion** [comment]
I'm sure there's no examples. If you not found tests for this, maybe I'd forgotten. Sorry, I'll take a look into this soon.

File: internal/sh/shell.go (hunk 2 above)
Comment (vitorarins) [inline internal/sh/shell.go]: Ah ok. Sorry for not seeing that.

**Comment by vitorarins** [comment]
No probs dude. I think testing this won't be trivial anyway. Thanks for the clear explanation. :wink:

**Comment by tiago4orion** [comment]
:x: [Build nash 1.0.0.68 failed](https://ci.appveyor.com/project/tiago4orion/nash/build/1.0.0.68) (commit https://github.com/NeowayLabs/nash/commit/4ad6b0fea8 by @tiago4orion)

**Comment by tiago4orion** [comment]
:x: [Build nash 1.0.0.71 failed](https://ci.appveyor.com/project/tiago4orion/nash/build/1.0.0.71) (commit https://github.com/NeowayLabs/nash/commit/10bf075f2d by @tiago4orion)

**Comment by tiago4orion** [comment]
@katcipis I merged windows branch here.. otherwise I cannot work on this anymore hahaha

**Comment by tiago4orion** [comment]
:x: [Build nash 1.0.0.83 failed](https://ci.appveyor.com/project/tiago4orion/nash/build/1.0.0.83) (commit https://github.com/NeowayLabs/nash/commit/50ef4097f0 by @tiago4orion)

**Comment by tiago4orion** [comment]
:x: [Build nash 1.0.0.85 failed](https://ci.appveyor.com/project/tiago4orion/nash/build/1.0.0.85) (commit https://github.com/NeowayLabs/nash/commit/1c41c4ee9f by @tiago4orion)

**Comment by tiago4orion** [comment]
:x: [Build nash 1.0.0.102 failed](https://ci.appveyor.com/project/tiago4orion/nash/build/1.0.0.102) (commit https://github.com/NeowayLabs/nash/commit/c734293aa3 by @tiago4orion)

**Comment by tiago4orion** [comment]
:x: [Build nash 1.0.0.118 failed](https://ci.appveyor.com/project/tiago4orion/nash/build/1.0.0.118) (commit https://github.com/NeowayLabs/nash/commit/12bdfb525f by @tiago4orion)

**Comment by tiago4orion** [comment]
:x: [Build nash 1.0.0.155 failed](https://ci.appveyor.com/project/tiago4orion/nash/build/1.0.0.155) (commit https://github.com/NeowayLabs/nash/commit/cc0882a9b1 by @tiago4orion)

**Comment by tiago4orion** [comment]
:x: [Build nash 1.0.0.157 failed](https://ci.appveyor.com/project/tiago4orion/nash/build/1.0.0.157) (commit https://github.com/NeowayLabs/nash/commit/e6723648d4 by @tiago4orion)

File: internal/sh/shell_test.go [hunk 7]
```diff
@@ -1,2386 +1,2433 @@
-package sh
```
Comment (tiago4orion) [inline internal/sh/shell_test.go]: :cry: :cry: :cry: :cry: :cry: :cry: :cry: :cry: :cry: :cry: :cry: :cry: :cry: :cry: :cry: :cry:

Comment (tiago4orion) [reply internal/sh/shell_test.go]: hmmm, this is the best way to put a backdoor in opensource projects :-P

Comment (katcipis) [reply internal/sh/shell_test.go]: I don't get it =(

Comment (tiago4orion) [reply internal/sh/shell_test.go]: because of the massive diff... because of the conversion from \n to \r... of stupid windows

Comment (tiago4orion) [reply internal/sh/shell_test.go]: easy to put malicious lines inside this kind of diffs... multiple lines that look identical.. but it's not. hehe

**Comment by tiago4orion** [comment]
:x: [Build nash 1.0.0.162 failed](https://ci.appveyor.com/project/tiago4orion/nash/build/1.0.0.162) (commit https://github.com/NeowayLabs/nash/commit/1e94a0dbc2 by @tiago4orion)

**Pull request merged**